src/main/java/com/hms_networks/sc/cumulocity/api/CConnectorMqttMgr.java

```diff
@@ -159,14 +164,47 @@ public void onMessage(MqttMessage mqttMessage) {
    * @param currentMqttStatus the current MQTT status integer
    */
   public void runOnMqttLoop(int currentMqttStatus) {
-    if (currentMqttStatus != MqttStatusCode.CONNECTED) {
-      // Increment loop attempt count
-      mqttLoopAttemptCount++;
-      if (mqttLoopAttemptCount > mqttLoopAttemptLimit) {
-        Logger.LOG_CRITICAL(
-            "MQTT disconnected for " + mqttLoopAttemptCount + " cycles, triggering a reprovision.");
-        CConnectorMain.rerunProvisioning();
+    // Retry pending payloads if connected to MQTT
+    if (currentMqttStatus == MqttStatusCode.CONNECTED) {
+      if (!pendingRetryMessages.empty()) {
+        // Get the first pending retry message
+        CConnectorRetryMessage retryPayload = (CConnectorRetryMessage) pendingRetryMessages.peek();
+
+        // Retry the payload and pop from stack if successful
+        try {
```

Comment:
The maximum number of pops and send is 1. Should this try for more. Is there any risk that the stack grows and never burns down when a connection is restored?

Reply:
If the connection is restored, there should be no possibility that the stack will grow (it should only shrink).

The reason for only trying one message at a time was to reduce the complexity of this code. It becomes increasingly more work if we want to control/constrain the iteration of a stack.

Reply:
If we only try one message, it is simple to handle success/failure with the retry. 

If we are iterating over the stack, for example, with `while (!pendingRetryMessages.empty())`, we could find ourselves in an infinite loop should messages fail to retry (because the stack will never be empty).

By only trying one message at a time, we are simplifying our code while guaranteeing we do not enter an unpredictable infinite loop condition.

Reply:
If for some reason, another exception prevents messages from being sent and they are queuing for retry, there is a maximum of 16 retries allowed per message. After that, the message is entirely dropped with an error message shown.